src/api/plugins/tdbp/warren_tdbp/indicators.py

```diff
@@ -0,0 +1,259 @@
+"""Warren TdBP indicators."""
+
+from datetime import datetime, timedelta
+from typing import List, Optional
+
+import pandas as pd
+from functools import cached_property
+from pydantic import HttpUrl
+from ralph.backends.http import LRSHTTP
+from ralph.backends.http.lrs import LRSQuery
+from ralph.conf import LRSHeaders
+from warren.backends import lrs_client as async_lrs_client
+from warren.filters import DatetimeRange
+from warren.indicators import BaseIndicator
+from warren.xi import ExperienceRead
+from warren.xi.clients import CRUDExperience as async_xi_experience_client
+
+from .models import ActiveAction, SlidingWindow, StudentScore
+
+
+class SlidingWindowIndicator(BaseIndicator):
+    """TODO"""
+
+    course_experiences: List[str] = None
+    sliding_window: DatetimeRange = None
+    active_actions: List[ActiveAction] = None
+    dynamic_cohort: List[str] = None
+
+    def __init__(  # noqa:PLR0913
+        self,
+        course_id: str,
+        until: Optional[datetime] = None,
+        sliding_window_size: int = 15,
+        active_actions_min: int = 6,
+        dynamic_cohort_size: int = 3,
+    ):
+        """Initializes sliding window indicator."""
+        self.course_id = course_id
+        self.sliding_window_size = sliding_window_size
+        self.active_actions_min = active_actions_min
+        self.dynamic_cohort_size = dynamic_cohort_size
+        self.until = until or datetime.now()
+
+    @property
+    async def xi_experience_client(self):
+        """To move to BaseIndicator in Warren core."""
+        return async_xi_experience_client
+
+    def get_xi_experience_query(self):
+        """To move to BaseIndicator in Warren core."""
+
+    def fetch_experiences(self) -> List[ExperienceRead]:
+        """Return actions related to course read from Experience Index."""
+
+        # TODO
+        # Parse xi_url with course_id
+        # Query experience index to get course_id experience
+        # Query for each retrieved related experience IDs and content info
+        # Return list of experience
+
+    def get_lrs_query(self, experience_id) -> LRSQuery:
+        """Get the LRS query for fetching required statements."""
+        return LRSQuery(
+            query={
+                "activity": experience_id,
+                "since": self.since.isoformat(),
```

Comment:
`self.since` will be empty as long as `self.span_range` is not defined, see: https://github.com/openfun/warren/blob/main/src/api/core/warren/indicators/base.py#L71-L74